specs/escrow.md

```diff
@@ -0,0 +1,60 @@
+
+The Sphinx team owns the `SphinxEscrow` contract.
+
+## Table of Contents
+
+- [Relevant files](#relevant-files)
+- [Use Case](#use-case)
+- [High-Level Invariants](#high-level-invariants)
+- [Function-Level Invariants](#function-level-invariants)
+- [Assumptions](#assumptions)
+
+## Relevant Files
+
+- The contract: [`SphinxEscrow.sol`](todo)
+- Unit tests: [`SphinxEscrow.t.sol`](todo)
+
+## Use Case
+
+### Allow Sphinx to accept crypto payments for the variable cost of deployments
+For Sphinx to accept crypto payments we must be able to guarantee that we will be paid for the cost of the deployment. Since the exact cost of a deployment cannot be exactly known ahead of time, we cannot charge for the deployment upfront. So instead, we estimate the cost of a deployment and require that the user transfer the estimated amount to an escrow contract while the deployment is being executed. After the deployment is completed, we calculate the exact cost and transfer the remaining funds back to the user.
+
+## High-Level Invariants
+- It must be possible to use the `SphinxEscrow` contract with any arbitrary ERC20 contract.
+- It must be possible to use the `SphinxEscrow` contract to transfer ERC20 tokens from an `owner` address to the `SphinxEscrow` contract if the `SphinxEscrow` contract has been granted an allowance.
+- It must *not* be possible to transfer ERC20 tokens from an `owner` address to any address other than the `SphinxEscrow` contract if the `SphinxEscrow` contract has been granted an allowance.
+- It must *only* be possible to transfer ERC20 tokens out of the `SphinxEscrow` contract to the original ERC20 `owner` address or an address with the `DEFAULT_ADMIN_ROLE` on the `SphinxEscrow` contract.
+
+## Function-Level Invariants
+
+#### `constructor(address _owner)`
+- Must grant the `DEFAULT_ADMIN_ROLE` to the specified owner address.
+- Must revert if the `_owner` address is `address(0)`.
+
+#### `function claim(address _owner, address _erc20, uint256 _amount)`
+- Must revert if `_amount` is greater than the allowance of the `SphinxEscrow` contract on `_erc20` for `_owner`.
+- Must revert if the `_owner` does not have a balance of `_erc20` that is >= `_amount`.
+- A successful call must:
+  - Transfer the `_amount` of `_erc20` to the `SphinxEscrow` contract.
+  - Increment the `escrowBalance` of `_erc20` for `_owner` by `_amount`.
+  - Emit a `Claimed` event in the `SphinxEscrow` contract.
+
+#### `function disburse(address _owner, address _feeRecipient, address _erc20, uint256 _refund, uint256 _cost)`
+- Must revert if the amount of `_erc20` currently in escrow for `_owner` is < `_refund` + `_cost`
```

Comment:
> - Must revert if the amount of `_erc20` currently in escrow for `_erc20` is < `_refund` + `_cost`

This invariant allows for the possibility that only a fraction of the escrowed `_erc20` is sent to the `_owner` or `_feeRecipient`. For example, if the escrowed `_erc20` amount is 20 USDC, and `_refund` and `_cost` are both 5, then an error isn't thrown by this function. Is that intentional? I could see this being useful if the user wants to keep a balance of ERC20 for all deployments instead of adding a new allowance for each deployment. However, this use case isn't mentioned in the "Use Case" section, so it's not clear what the expected behavior is